### PR TITLE
Insert to table with multi primary or without autoicrement primary

### LINKED ISSUE #41: Bug: Return value in Database::table()->insert() without autoincrement on PK
https://github.com/nette/database/blob/master/src/Database/Table/Selection.php#L742

$primaryKey = $this->context->getInsertId() aka PDO::lastInsertId()

you have table:
id int(10) unsigned NOT NULL
value varchar(255) COLLATE utf8_czech_ci DEFAULT NULL
PRIMARY KEY (`id`),

Database::table()->insert(array('id'=>2, 'value'=> 'text'));
will result in
PDO::lastInsertId() return 0 because we have not autoincrement on the table
https://github.com/nette/database/blob/master/src/Database/Table/Selection.php#L769

```
$row = $this->createSelectionInstance()
            ->select('*')
            ->wherePrimary(0)
            ->fetch();
```

solution?
check no autoincrement on PK and then iterate over data to insert and get primary key value from this array?

### PR DESCRIPTION
**Pull request by ricco24**
Insert to table without autoincrement primary or with multiple column primary returns inserted record.

I cant find bug1342.postgre but i think this solve this bug too. Maybe i can delete this bug report test file.

fixes #41 #80

### CODE REVIEW

**Comment by Unlink** [comment]
Duplicate #102?
My PR also handles situations, where you have composite primary key and one part of it has autoincrement.
Try this:
https://github.com/nette/database/pull/102/files#diff-a5ef18703e35a05b08eb6089336ad773R21

**Comment by ricco24** [comment]
Yes, it is duplicate. I noticed your PR until after I made ​​my. So i let it here.
But this PR handles composite key with one autoincrement part too.

**Comment by dg** [comment]
So ... which way to go? ;-)

**Comment by Unlink** [comment]
@dg I'm not sure, i don't have completed API for IStructrure in my PR and i found there some issues.
so if @ricco24 adjusts them, i'll close my PR.

**Comment by mikoczy** [comment]
:+1:

**Comment by mikoczy** [comment]
this could fix another problem, which occured in php 7.1 (pdo, pgsql driver).. (specifically [FR #72633](https://bugs.php.net/bug.php?id=72633)) this means, that the lastInsertId method could return a value from a different table than the current one, if that row was inserted in the same database session. For example if we insert a row into a table with an auto increment primary key, pdo-s lastInsertId will return the correct number. But if we insert another row into a different table in the same database session (with uuid primary key), lastInsertId will still return the numeric id from the previous query. This behaviour causes an error in the current implementation of the insert method in Nette\Database\Table\Selection, because after insertion, it tries to fetch a row (from a table with uuid primary key) by a numeric ID, which was inserted into a different table previously...

**Comment by tomaj** [comment]
👍

**Comment by ricco24** [comment]
@dg what with this? It is possible to merge or not? In current state nette database is unusable with php7.1, postgres 8.3+ and uuid as priamry key. Every insert fails ...

**Comment by dg** [comment]
This may cause a BC break? (Except that it adds new method to interface.) Or is it suitable for 2.4?

File: tests/Database/files/mysql-nette_test4.sql [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/*!40102 SET storage_engine = InnoDB */;
+/*!40102 SET default_storage_engine = InnoDB */;
```
Comment (dg) [inline tests/Database/files/mysql-nette_test4.sql]: What about `CREATE TABLE ... ENGINE=InnoDB` as used in other files?

File: src/Database/Structure.php [hunk 2]
```diff
@@ -65,6 +65,33 @@ public function getPrimaryKey($table)
...
+	{
+		$primaryKey = $this->getPrimaryKey($table);
+		if (!$primaryKey) {
+			return NULL;
+		}
+
+		// Search for autoincrement key from multi primary key
+		if (is_array($primaryKey)) {
+			foreach ($this->getColumns($table) as $column) {
+				if (in_array($column['name'], $primaryKey) && $column['autoincrement']) {
```
Comment (dg) [inline src/Database/Structure.php]: in_array() is relative slow operation, is possible this?

```php
		if (is_array($primaryKey)) {
			$keys = array_flip($primaryKey);
			foreach ($this->getColumns($table) as $column) {
				if (isset($keys[$column['name']]) && $column['autoincrement']) {
					return $column['name'];
				}
			}
			return NULL;
		}
```

**Comment by dg** [comment]
What about move searching for keys from `getPrimaryAutoincrementKey` (and maybe from `getPrimaryKeySequence` too) to `loadStructure` to use cache?

**Comment by ricco24** [comment]
That seems to be a good idea. I can make another pr with optimized loadStructure() function when this will be merged.

**Comment by dg** [comment]
Ok, so I'll merge it.

**Pull request merged**